docs/architecture/engine.md

```diff
@@ -0,0 +1,197 @@
+2) The in-memory cached block (also located inside an internal lookup in the Series object) **[4PM - 6PM]**
+3) The block from disk (the block retrieve from disk will then be cached according to the current [caching policy](caching.md) **[2PM - 4PM]**
+
+Retrieving blocks from the active buffers and in-memory cache is simple, the data is already present in memory and easily accessible via hashmaps keyed by series ID. Retrieving a block from disk is more complicated. The flow for retrieving a block from disk is as follows:
+
+1. Consult the in-memory bloom filter to determine if its possible the series exists on disk.
+2. If the bloom filter returns positive, then binary search the in-memory index summaries to find the nearest index entry that is *before* the series ID that we're searching for. Review the `index_lookup.go` file for implementation details.
+3. Jump to the offset in the index file that we obtained from the binary search in the previous step, and begin scanning forward until we identify the index entry for the series ID we're looking for *or* we get far enough in the index file that it becomes clear that the ID we're looking for doesn't exist (this is possible because the index file is sorted by ID)
+4. Jump to the offset in the data file that we obtained from scanning the index file in the previous step, and begin streaming data.
+
+
+Once M3DB has retrieved the three blocks from their respective locations in memory / on-disk, it will transmit all of the data back to the client. 
+
+**Note:** Since M3DB nodes return compressed blocks (the M3DB client decompresses them) its not possible to return "partial results" for a given block. If any portion of a read requests spans a given block, then that block in its entirety must be transmitted back to the client. In practice, this ends up being not much of an issue because of the high compression ratio that M3DB is able to achieve.
+
+## Background processes
+
+M3DB has a variety of processes that run in the background during normal operation.
+
+### Ticking
+
+The ticking process runs continously in the background and is responsible for a variety of tasks:
+
+1. Merging duplicate encoders for a given series / block start combination
+2. Removing expired / flushed series and blocks from memory
+
+
+#### Merging duplicate encoders
+
+M3TSZ is designed for compressing time series data in which each datapoint has a timestamp that is larger than the last encoded datapoint. For monitoring workloads this works very well because every subsequent datapoint is almost always larger than the previous one. However, real world systems are messy and occassionally out of order writes will be received. When this happens, M3DB will allocate a new encoder for the out of order datapoints. The duplicate encoders need to be merged before flushing the data to disk, but to prevent huge memory spikes during the flushing process we continuously merge out of order encoders in the background.
```

Comment:
should this also be mentioned in the write path?

Reply:
it is

```
There is also some special logic for handling out-of-order writes which is discussed in the [merging all encoders section](engine.md#merging-all-enoders).
```

Reply:
dope